import os
from Bio import SeqIO

##### Configuration #####

config["analysis_stem"]=[i for i in config["analysis_stem"].split(',')]

##### Target rules #####

rule all:
    input:
        expand(config["output_path"] + "/consensus_sequences/{sample}.fasta",sample=config["sample"]),
        expand(config["output_path"] + "/binned_{sample}/polishing/{analysis_stem}/mapped.paf",sample=config["sample"],analysis_stem=config["analysis_stem"])

##### Modules #####
    
include: "rules/map_polish.smk"

rule trim_primers:
    input:
        ref = config["output_path"] + "/binned_{sample}/{analysis_stem}.fasta",
        reads = config["output_path"]+"/binned_{sample}/{analysis_stem}.fastq",
        csv = config["output_path"] + "/binned_{sample}/{analysis_stem}.csv",
        amp_file = config["amplicon_file"]
    params:
        path_to_script = 
    output:
        config["output_path"]+"/binned_{sample}/{analysis_stem}.primer_trimmed.fastq"
    shell:

    
        





